### PR TITLE
[YUNIKORN-1090] remove common.Node from the shim

### PR DESCRIPTION
**Pull request by manirajv06**
### What is this PR for?

Enhanced si_helper methods to create/update/delete node request and use these methods directly instead of using common.Node method wrappers to avoid the unnecessary object creations. Also re-used the methods in other place as well.

### What type of PR is it?
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1090

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/cache/nodes.go [hunk 1]
```diff
@@ -213,21 +212,22 @@ func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
 		return
 	}
 
-	if schedulerNode, ok := nc.nodesMap[newNode.Name]; ok {
-		schedulerNode.ready = common.HasReadyCondition(newNode)
+	schedulerNode, ok := nc.nodesMap[newNode.Name]
+	if ok {
+		schedulerNode.ready = hasReadyCondition(newNode)
```
Comment (pbacsko) [inline pkg/cache/nodes.go:217]: The logic here changed slightly. We only send out the update to the proxy if the node is found in the `nodesMap`. I don't think it has substantial effects, just wondering if this is better?

cc @wilfred-s

**Comment by pbacsko** [comment]
Mostly LGTM, I added two minor comments.

File: pkg/common/si_helper.go [hunk 2]
```diff
@@ -177,14 +184,11 @@ func CreateUpdateRequestForUpdatedNode(node Node) si.NodeRequest {
 	return request
 }
 
-func CreateUpdateRequestForDeleteNode(node Node) si.NodeRequest {
+func CreateUpdateRequestForDeleteNode(nodeID string, action si.NodeInfo_ActionFromRM) si.NodeRequest {
```
[the remark this answers is not in the document]
Comment (manirajv06) [reply pkg/common/si_helper.go]: "delete" as generic term for drain or decommission, but it does restore as well. hence named it as DeleteOrRestore. Hope this is ok.

File: pkg/cache/nodes.go (hunk 1 above)
Comment (manirajv06) [reply pkg/cache/nodes.go:217]: Yes, it should not have any effects. I don't see a situation wherein entry doesn't exist in `nodesMap `during "update"

File: pkg/cache/nodes.go [hunk 3]
```diff
@@ -213,21 +212,22 @@ func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
 		return
```
Comment (wilfred-s) [inline pkg/cache/nodes.go:212]: we must update the resources in the cachedNode, might need locking and a setter method:
`cachedNode.capacity = common.GetNodeResource(&newNode.Status)`

File: pkg/cache/nodes.go (hunk 1 above)
Comment (wilfred-s) [reply pkg/cache/nodes.go:217]: I do think we need to carefully look at what we do here (referencing new line numbers):
* line 195: we get the node out of the nodesMap store in cachedNode
* line 215: we retrieve the same node again, we already know it CANNOT be nil and ok is `true, we should re-use cachedNode
* line 217: should always be executed, even if resources do not change, should this use a locked setter method?

File: pkg/cache/nodes.go [hunk 4]
```diff
@@ -213,21 +212,22 @@ func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
...
 	}
 
-	if schedulerNode, ok := nc.nodesMap[newNode.Name]; ok {
-		schedulerNode.ready = common.HasReadyCondition(newNode)
+	schedulerNode, ok := nc.nodesMap[newNode.Name]
+	if ok {
+		schedulerNode.ready = hasReadyCondition(newNode)
 		if !schedulerNode.ready {
 			log.Logger().Debug("Node is not ready", zap.String("Node name", newNode.Name))
 		}
```
Comment (wilfred-s) [inline pkg/cache/nodes.go:220]: we can log the state of the node always

File: pkg/cache/nodes.go [hunk 5]
```diff
@@ -213,21 +212,22 @@ func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
...
+		request := common.CreateUpdateRequestForUpdatedNode(newNode.Name, common.GetNodeResource(&newNode.Status),
```
Comment (wilfred-s) [inline pkg/cache/nodes.go:221]: should use the cachedNode, does this event need to be generated when ready changes without resource changes?

File: pkg/cache/nodes.go [hunk 6]
```diff
@@ -213,21 +212,22 @@ func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
...
+			schedulerNode.occupied, schedulerNode.ready)
+		log.Logger().Info("report updated nodes to scheduler", zap.Any("request", request))
+		if err := nc.proxy.UpdateNode(&request); err != nil {
+			log.Logger().Info("hitting error while handling UpdateNode", zap.Error(err))
+		}
 	} else {
 		log.Logger().Error("Unable to find scheduler node in nodes map", zap.String("node name",
```
Comment (wilfred-s) [inline pkg/cache/nodes.go:228]: can be removed node is always found

File: pkg/cache/nodes.go [hunk 7]
```diff
@@ -259,3 +258,12 @@ func (nc *schedulerNodes) schedulerNodeEventHandler() func(obj interface{}) {
 		}
 	}
 }
+
+func hasReadyCondition(node *v1.Node) bool {
+	for _, condition := range node.Status.Conditions {
```
Comment (wilfred-s) [inline pkg/cache/nodes.go:263]: Can `node` or `Status` be nil? Rather be careful than rely on the externally defined object always being complete.

File: pkg/common/si_helper.go [hunk 8]
```diff
@@ -133,18 +134,23 @@ func CreateReleaseAllocationRequestForTask(appID, allocUUID, partition, terminat
 	return result
 }
 
-func CreateUpdateRequestForNewNode(node Node) si.NodeRequest {
+func CreateUpdateRequestForNewNode(nodeID string, capacity *si.Resource, occupied *si.Resource,
```
Comment (wilfred-s) [inline pkg/common/si_helper.go:137]: Add comment to the exported function:
```
// CreateUpdateRequestForNewNode builds a NodeRequest ....
```
Should do the same for all si_helper functions changed here.

File: pkg/common/si_helper.go [hunk 9]
```diff
@@ -133,18 +134,23 @@ func CreateReleaseAllocationRequestForTask(appID, allocUUID, partition, terminat
...
+	existingAllocations []*si.Allocation, labels string, ready bool) si.NodeRequest {
 	// Use node's name as the NodeID, this is because when bind pod to node,
 	// name of node is required but uid is optional.
 	nodeInfo := &si.NodeInfo{
-		NodeID:              node.name,
-		SchedulableResource: node.capacity,
+		NodeID:              nodeID,
+		SchedulableResource: capacity,
+		OccupiedResource:    occupied,
 		// TODO is this required?
```
Comment (wilfred-s) [inline pkg/common/si_helper.go:145]: Please remove this TODO

File: pkg/common/si_helper.go [hunk 10]
```diff
@@ -156,15 +162,16 @@ func CreateUpdateRequestForNewNode(node Node) si.NodeRequest {
 	return request
 }
 
-func CreateUpdateRequestForUpdatedNode(node Node) si.NodeRequest {
+func CreateUpdateRequestForUpdatedNode(nodeID string, capacity *si.Resource, occupied *si.Resource,
+	ready bool) si.NodeRequest {
 	// Currently only includes resource in the update request
```
Comment (wilfred-s) [inline pkg/common/si_helper.go:167]: Please update this comment as it is incorrect

File: pkg/common/si_helper_test.go [hunk 11]
```diff
@@ -18,14 +18,21 @@
...
 	"testing"
 
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/constants"
 	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/common"
+	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/si"
+
 	"gotest.tools/assert"
+
 	v1 "k8s.io/api/core/v1"
 	apis "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (wilfred-s) [inline pkg/common/si_helper_test.go:31]: Can we re-order as we are cleaning up, 3 groups separated by an empty line:
- group system imports
- group other imports
- group yunikorn imports

**Review by wilfred-s — CHANGES_REQUESTED** [review]
One other main point for the `updateNode`:
* line 203-208: if we need to drain or restore we send an update to the core to do that and keep processing: we again look up the node during the calls in the nodesMap (WHY ???) and do some checks. That needs to be optimised out and simplified (remove functions drainNode and restoreNode) We also do not need an `if..else` construct as the checks can never be true at the same time:
```
	if !oldNode.Spec.Unschedulable && newNode.Spec.Unschedulable {
		&& cachedNode.getNodeState() == events.States().Node.Healthy {
			dispatcher.Dispatch(CachedSchedulerNodeEvent{
				NodeID: cachedNode.name,
				Event:  events.DrainNode,
			})
	}
	if oldNode.Spec.Unschedulable && !newNode.Spec.Unschedulable
		&& cachedNode.getNodeState() == events.States().Node.Draining {
			dispatcher.Dispatch(CachedSchedulerNodeEvent{
				NodeID: cachedNode.name,
				Event:  events.RestoreNode,
			})
	}
```

File: pkg/cache/nodes.go (hunk 3 above)
Comment (manirajv06) [reply pkg/cache/nodes.go:186]: Taken care here itself even though YUNIKORN-1091 has been filed for the same

**Comment by manirajv06** [comment]
Observed few more things while doing this clean up:

1. I can see few more places where SchedulerNode properties gets updates without having lock mechanism in place. For example, `schedulerNode.occupied = common.Add(schedulerNode.occupied, resource)` in `updateNodeOccupiedResources` method
2. I think we should start exploring and see if `sync.Map` makes sense instead of using `map` given the no. of reads and writes it is going through.

Please share your views.

**Comment by wilfred-s** [comment]
> Observed few more things while doing the clean up:
> 
> 1. I can see few more places where SchedulerNode properties gets updated without having proper lock mechanism in place. For example, `schedulerNode.occupied = common.Add(schedulerNode.occupied, resource)` in `updateNodeOccupiedResources` method. We should call `setOccupiedResource` here.

We should look at all those and we might need to get this fixed throughout. A follow up jira would be the right approach here.

> 2. I think we should start exploring and see if `sync.Map` makes sense instead of using `map` given the no. of reads and writes it is going through.

The number of writes is minimal. Node updates should not happen at a high rate, not even in large clusters. Reads do occur more often. Optimising the performance of this access would most likely not affect the overall performance. `sync.Map` also has a different behaviour when ranging over entries which could cause small but really difficult to track down bugs.

> 3. Among multiple conditions discussed in https://kubernetes.io/docs/concepts/architecture/nodes/#condition, only "Ready" has been used. Should we use other conditions as well to determine a generic `isNodeHealthy` factor and eventually passing to core rather than depending on only one condition?

That is surely worth considering. We should file a follow up jira and assess what we can do in this area.

**Review by wilfred-s — APPROVED** [review]
LGTM
the other points mentioned we should fix via follow up jiras

**Pull request closed without merging**